### ISSUE TITLE
Fast Track II ASIO error

### ISSUE THREAD

**Issue by bobovuk** [mdias/rs_asio#396]
Hello, I have an old interface M-Audio Fast Track II that worked with NoCableLauncher and Rocksmith 2014. I was hoping I could also get it working with RS_ASIO, but I am getting errors initializing the interface. It's a fresh Steam installation of Rocksmith 2014 and here is the log:

```
0.000 [INFO]   - Wrapper DLL loaded (v0.7.0)
0.000 [INFO]  PatchOriginalCode
0.032 [INFO]  image crc32: 0xd1b38fcb
0.145 [INFO]  Patching CoCreateInstance
0.145 [INFO]  Patch_CallAbsoluteIndirectAddress - num locations: 2
0.145 [INFO]  Patching call at 00C02DBD
0.145 [INFO]  Patching call at 00E75278
0.145 [INFO]  Patching PortAudio MarshalStreamComPointers
0.145 [INFO]  Patch_CallRelativeAddress - num locations: 1
0.145 [INFO]  Patching call at 00E7483F
0.145 [INFO]  Patching PortAudio UnmarshalStreamComPointers
0.145 [INFO]  Patch_CallRelativeAddress - num locations: 1
0.145 [INFO]  Patching call at 00E748F4
0.145 [INFO]  Patching Two Guitar Tones Connected Message Box (num locations: 1)
0.145 [INFO]  Patching bytes at 007C0C5F
0.313 [INFO]  Patched_CoCreateInstance called: {DC12A687-737F-11CF-884D-00AA004B2E24}
0.485 [INFO]  Patched_CoCreateInstance called: {DC12A687-737F-11CF-884D-00AA004B2E24}
0.715 [INFO]  Patched_CoCreateInstance called: {DC12A687-737F-11CF-884D-00AA004B2E24}
0.959 [INFO]  Patched_CoCreateInstance called: {DC12A687-737F-11CF-884D-00AA004B2E24}
1.164 [INFO]  Patched_CoCreateInstance called: {DC12A687-737F-11CF-884D-00AA004B2E24}
1.343 [INFO]  Patched_CoCreateInstance called: {DC12A687-737F-11CF-884D-00AA004B2E24}
1.514 [INFO]  Patched_CoCreateInstance called: {DC12A687-737F-11CF-884D-00AA004B2E24}
1.741 [INFO]  Patched_CoCreateInstance called: {DC12A687-737F-11CF-884D-00AA004B2E24}
2.056 [INFO]  Patched_CoCreateInstance called: {DC12A687-737F-11CF-884D-00AA004B2E24}
2.341 [INFO]  Patched_CoCreateInstance called: {DC12A687-737F-11CF-884D-00AA004B2E24}
2.614 [INFO]  Patched_CoCreateInstance called: {DC12A687-737F-11CF-884D-00AA004B2E24}
2.933 [INFO]  Patched_CoCreateInstance called: {DC12A687-737F-11CF-884D-00AA004B2E24}
3.153 [INFO]  Patched_CoCreateInstance called: {DC12A687-737F-11CF-884D-00AA004B2E24}
3.402 [INFO]  Patched_CoCreateInstance called: {DC12A687-737F-11CF-884D-00AA004B2E24}
3.727 [INFO]  Patched_CoCreateInstance called: {DC12A687-737F-11CF-884D-00AA004B2E24}
4.125 [INFO]  Patched_CoCreateInstance called: {DC12A687-737F-11CF-884D-00AA004B2E24}
4.430 [INFO]  Patched_CoCreateInstance called: {DC12A687-737F-11CF-884D-00AA004B2E24}
4.747 [INFO]  Patched_CoCreateInstance called: {DC12A687-737F-11CF-884D-00AA004B2E24}
5.006 [INFO]  Patched_CoCreateInstance called: {DC12A687-737F-11CF-884D-00AA004B2E24}
5.263 [INFO]  Patched_CoCreateInstance called: {DC12A687-737F-11CF-884D-00AA004B2E24}
5.557 [INFO]  Patched_CoCreateInstance called: {DC12A687-737F-11CF-884D-00AA004B2E24}
5.886 [INFO]  Patched_CoCreateInstance called: {DC12A687-737F-11CF-884D-00AA004B2E24}
6.730 [INFO]  Patched_CoCreateInstance called: IID_IMMDeviceEnumerator
6.731 [INFO]  DebugDeviceEnum::GetDefaultAudioEndpoint - dataFlow: eRender - role: eMultimedia
6.731 [INFO]  RSAggregatorDeviceEnum::UpdateAvailableDevices
6.731 [INFO]  GetWineAsioInfo - Looking for wineasio.dll... 
6.732 [INFO]    Failed to load wineasio.dll or file not found
6.732 [INFO]  AsioHelpers::FindDrivers
6.732 [INFO]    Fast Track ASIO
6.732 [INFO]  RSAsioDeviceEnum::UpdateAvailableDevices - output requesting ASIO driver: Fast Track ASIO (64-bit)
6.732 [ERROR]  RSAsioDeviceEnum::UpdateAvailableDevices - failed.
6.732 [INFO]  RSAsioDeviceEnum::UpdateAvailableDevices - input[1] requesting ASIO driver: Fast Track ASIO (64-bit)
6.732 [ERROR]  RSAsioDeviceEnum::UpdateAvailableDevices - failed.
6.732 [INFO]  DebugDeviceEnum::UpdateAvailableDevices - 0 render devices, 0 capture devices
6.733 [INFO]    hr: 80070490
6.733 [INFO]    *ppEndpoint: 00000000
6.733 [INFO]  DebugDeviceEnum::GetDefaultAudioEndpoint - dataFlow: eCapture - role: eMultimedia
6.733 [INFO]    hr: 80070490
6.733 [INFO]    *ppEndpoint: 00000000
6.733 [INFO]  DebugDeviceEnum::EnumAudioEndpoints - dataFlow: eAll - dwStateMask: 1
6.733 [INFO]    hr: S_OK
6.733 [INFO]    *ppDevices: 1343A358
12.498 [INFO]  Patched_CoCreateInstance called: IID_IMMDeviceEnumerator
12.498 [INFO]  DebugDeviceEnum::GetDefaultAudioEndpoint - dataFlow: eRender - role: eMultimedia
12.498 [INFO]  RSAggregatorDeviceEnum::UpdateAvailableDevices
12.498 [INFO]  AsioHelpers::FindDrivers
12.498 [INFO]    Fast Track ASIO
12.498 [INFO]  RSAsioDeviceEnum::UpdateAvailableDevices - output requesting ASIO driver: Fast Track ASIO (64-bit)
12.498 [ERROR]  RSAsioDeviceEnum::UpdateAvailableDevices - failed.
12.498 [INFO]  RSAsioDeviceEnum::UpdateAvailableDevices - input[1] requesting ASIO driver: Fast Track ASIO (64-bit)
12.498 [ERROR]  RSAsioDeviceEnum::UpdateAvailableDevices - failed.
12.498 [INFO]  DebugDeviceEnum::UpdateAvailableDevices - 0 render devices, 0 capture devices
12.498 [INFO]    hr: 80070490
12.498 [INFO]    *ppEndpoint: 00000000
12.498 [INFO]  DebugDeviceEnum::GetDefaultAudioEndpoint - dataFlow: eCapture - role: eMultimedia
12.498 [INFO]    hr: 80070490
12.498 [INFO]    *ppEndpoint: 00000000
12.498 [INFO]  DebugDeviceEnum::EnumAudioEndpoints - dataFlow: eAll - dwStateMask: 1
12.498 [INFO]    hr: S_OK
12.498 [INFO]    *ppDevices: 1343A8F8
12.505 [INFO]  DebugDeviceEnum::RegisterEndpointNotificationCallback
15.716 [INFO]  DebugDeviceEnum::UnregisterEndpointNotificationCallback
15.898 [INFO]   - Wrapper DLL unloaded 
```
I am using the Mod Installer/DLL Modifier 1.2.7.4 to make config changes. Here is the RS-ASIO.ini:

```
[Config]
EnableWasapiOutputs=0
EnableWasapiInputs=0
EnableAsio=1

[Asio]
; available buffer size modes:
;    driver - respect buffer size setting set in the driver
;    host   - use a buffer size as close as possible as that requested by the host application
;    custom - use the buffer size specified in CustomBufferSize field
BufferSizeMode=custom
CustomBufferSize=96

[Asio.Output]
Driver=Fast Track ASIO (64-bit)
BaseChannel=0
AltBaseChannel=
EnableSoftwareEndpointVolumeControl=1
EnableSoftwareMasterVolumeControl=1
SoftwareMasterVolumePercent=100
EnableRefCountHack=

[Asio.Input.0]
Driver=
Channel=1
EnableSoftwareEndpointVolumeControl=1
EnableSoftwareMasterVolumeControl=1
SoftwareMasterVolumePercent=100
EnableRefCountHack=

[Asio.Input.1]
Driver=Fast Track ASIO (64-bit)
Channel=1
EnableSoftwareEndpointVolumeControl=1
EnableSoftwareMasterVolumeControl=1
SoftwareMasterVolumePercent=100
EnableRefCountHack=

[Asio.Input.Mic]
Driver=
Channel=1
EnableSoftwareEndpointVolumeControl=1
EnableSoftwareMasterVolumeControl=1
SoftwareMasterVolumePercent=100
EnableRefCountHack=
```

Rocksmith throws a 'no output device detected' error on startup. If I select the WASAPi output, it launches with output working (using board audio) but the input still doesn't work, I assume because it can't initialize the Fast Track interface. I have set 48khz on the interface control panel, and 96 buffer on both the interface settings and the RS-ASIO settings.

Any ideas how I could troubleshoot, other than buying a new interface?
Thanks,
bobovuk

**Comment by mdias** [comment]
RS ASIO found the `Fast Track ASIO` driver but you're trying to use `Fast Track ASIO (64-bit)`, why?
Rocksmith is a 32-bit application, it will not load 64-bit drivers.

**Comment by bobovuk** [comment]
Unfortunately, the final driver for this old interface is only available as 64-bit, or at least I could not find the 32-bit version. In the meantime, I accepted the fact that I had to let go of this old interface and have since upgraded to Presonus Studio 26c, which of course works perfectly fine with the Rocksmith setup. 

![image](https://github.com/mdias/rs_asio/assets/150103554/591d071d-3c7b-48d8-9aac-5c0739dd9a87)

Thanks for replying.
